tools/slang-capability-generator/capability-generator-main.cpp

```diff
@@ -259,11 +421,82 @@ struct CapabilityConjunction
     }
 };
 
+String tryGetCapabilityName(const CapabilityDef* def)
+{
+    if(def)
+        return def->name;
+    return "[none]";
+}
+
 struct CapabilityDisjunction
 {
     List<CapabilityConjunction> conjunctions;
 
-    void addConjunction(const CapabilityConjunction& c)
+    void addConjunction(DiagnosticSink* sink, SourceLoc sourceLoc, const CapabilityConjunction& c)
+    {
+        if (c.isImpossible())
+            return;
+        bool cImpliesThis = false;
+        for (Index i = 0; i < conjunctions.getCount();)
+        {
+            // implied sets will be replaced
+            if (c.implies(conjunctions[i]))
+            {
+                cImpliesThis = true;
+                conjunctions.fastRemoveAt(i);
+            }
+            else
+                i++;
+        }
+        if (cImpliesThis)
+        {
+            conjunctions.add(_Move(c));
+            return;
+        }
+
+        for (Index i = 0; i < conjunctions.getCount();)
+        {
+            if (conjunctions[i].implies(c))
+            {
+                // subset is implied, we do not need to add it.
+                return;
+            }
+            else
+            {
+                // validate we are not creating a disjunction of same targets
+                if (conjunctions[i].shareTargetAndStageAtom(c))
+                {
+                    if (sink)
+                    {
+                        sink->diagnose(sourceLoc, Diagnostics::unionWithSameAbstractAtomButNotSubset, conjunctions[i].toString(), c.toString());
+                        sink = nullptr;
+                    }
+                }
+                i++;
+            }
+        }
+        conjunctions.add(_Move(c));
+    }
+    void removeImplied()
+    {
+        for (Index i = 0; i < conjunctions.getCount(); i++)
+        {
+            for (Index ii = 0; ii < conjunctions.getCount(); ii++)
```

Comment:
Can ii start from i+1?
I think `a.implies(b)` is same as `b.implies(a)`, isn't it?

Reply:
I meant that you can change the code to do the following,
```
for (Index ii = i+1; ii < conjunctions.getCount(); ii++)
```

Reply:
the code suggestion is incorrect: `conjunctions[x].implies(conjunctions[y]) != conjunctions[y].implies(conjunctions[x])`

Reply:
Please avoid using symbols when it can be simpler in a plan English.
In this specific case, your statement/assertion is not true always.
There surely will be a case where the following condition is true,
```
conjunctions[x].implies(conjunctions[y]) == conjunctions[y].implies(conjunctions[x])
```
Please try to avoid miscommunication more.

Reply:
The full loops here are correct. Both i and ii needs to be from 0 to the end, because `a.implies(b)` is not equivalent to `b.implies(a)`. We need to make sure both (a,b) and (b,a) are tested.

The algebraic implication operation is not commutative.

Reply:
I will reword the following:

the code suggestion is not going to function as intended: 
`conjunctions[x].implies(conjunctions[y])` is only equal to `conjunctions[y].implies(conjunctions[x])` when `y == x`. Implication is not a commutative operation, we cannot make this change.